sync.js

```diff
@@ -20,181 +17,120 @@ const SYNCFILE_FORMATS = {
+  return peer
+}
+
+function FileTarget (filename) {
+  return {
+    id: path.basename(filename),
+    filename,
+    type: 'file'
+  }
+}
+
 class Sync extends events.EventEmitter {
   constructor (osm, media, opts) {
     super()
-    if (!opts) opts = {}
+    this.opts = Object.assign(DEFAULT_OPTS, opts)
     opts.writeFormat = opts.writeFormat || 'hyperlog-sneakernet'
     if (!SYNCFILE_FORMATS[opts.writeFormat]) throw new Error('unknown syncfile write format: ' + opts.writeFormat)
 
     this.osm = osm
     this.media = media
-    this.host = opts.host
-    this.id = opts.id || 'Mapeo_' + randombytes(8).toString('hex')
+    this.id = opts.id || randombytes(8).toString('hex')
     this.opts = opts
+
+    // track replication progress states of files and wifi streams
     this._targets = {}
-    this._replicationServer = null
+
+    this.swarm = null
+  }
+
+  _swarm () {
+    this.swarm = Swarm()
+    this.swarm.on('connection-closed', (connection, info) => {
+      const target = WifiTarget(info)
+      this.emit('down', target)
+      debug('down', target)
+      delete this._targets[target.id]
+    })
+
+    this.swarm.on('connection', (connection, info) => {
+      // Skip your own machine.
+      if (info.id === this.id) {
+        debug('skipping sync target: it\'s me')
+        return
+      }
+
+      const target = WifiTarget(info)
+      this._targets[target.id] = target
+      this.emit('connection', target)
+      debug('connection', target)
+    })
     this._onerror = this._onerror.bind(this)
   }
 
   targets () {
     return values(this._targets)
   }
 
-  /**
-   * Sync to Target
-   * @type {[type]}
-   */
   syncToTarget (_target, opts) {
-    var self = this
-    if (!opts) opts = {}
-
-    const emitter = new events.EventEmitter()
-    const url = `ws://${_target.host}:${_target.port}`
-
-    var pending = 2
-    var target = this._targets[_target.name]
+  }
 
-    const done = (err) => {
-      if (err) {
-        if (target) {
-          target.status = 'replication-error'
-          target.message = err.message
-        }
-        return emitter.emit('error', err)
-      }
-      if (--pending === 0) {
-        if (target) {
-          target.status = 'replication-complete'
-        }
-        emitter.emit('end')
-      }
-    }
-    emitter.emit('progress', 'replication-started')
-    if (target) target.status = 'replication-started'
+  _syncError (err, target) {
+    target.status = 'replication-error'
+    target.message = err.message
+  }
 
-    const ws = wsock(`${url}/osm`, {
-      perMessageDeflate: false, objectMode: true
-    })
-    // TODO: real progress events
-    ws.once('data', () => {
-      if (target) target.status = 'osm-connected'
-      emitter.emit('progress', 'osm-connected')
-    })
+  _syncComplete (target) {
+    target.status = 'replication-complete'
+  }
 
-    const osm = self.osmReplicationStream(opts.osm)
-    replicate(osm, ws, done)
+  // FIXME: think about this API more
```

Comment:
I haven't thought at all about how to handle the structure here for `media-progress` and `osm-progress`  etc @noffle any ideas?

Probably would want to connect the frontend via a websocket to `mapeo-server`, which grabs the state and pushes it to the frontend every X milliseconds